### PR TITLE
Done, feature recreate

### LINKED ISSUE #787: deduplication does not work after migration from attic
From the mailing list:
```
I was able to reproduce the problem on a clean repo and small amount of
static data:

attic init repo
attic create --stats repo::backup1 data
borg upgrade repo
rm -rf ~/.cache/borg
borg create -v --stats repo::backup2 data (deduplication doesn't work)
rm -rf ~/.cache/borg
borg create -v --stats repo::backup3 data (deduplication does work)

Also I tried a number of actions with repository to recreate cache
(borg prune for example or borg create -n), but it didn't help.
```

### PR DESCRIPTION
**Pull request by enkore**
Use with caution: permanent data loss by specifying incorrect patterns/PATHs
is easily possible. Make a dry run to make sure you got everything right.

borg rewrite has many uses:
- Can selectively remove files/dirs from old archives, e.g. to free
  space or purging picturarum biggus dickus from history
- Recompress data
- Rechunkify data, to have upgraded Attic / Borg 0.xx archives deduplicate
  with Borg 1.x archives. (Or to experiment with chunker-params for
  specific use cases

It is interrupt- and resumable. (And for large files it will resume inside a file, so very nice for things like VM/disk images etc)

Chunks are not freed on-the-fly.
Rationale:
  Makes only sense when rechunkifying, but logic on which new chunks to
  free what input chunks is complicated and *very* delicate.

- Renamed code/feature/docs to "recreate", rationale:
  - While rewrite has the nice notion of "rewriting history" that's only one use case
  - Users could think that stuff is modified on the fly, which isn't the case
    - rewrite/recreate makes the illusion that archives are mutable: they are not. a new archive is created, and then the old one is deleted. as before, the fingerprint/ID of an archive forms a hash-tree and _always_ changes when the archive is changed in _any_ way.
  - "rewrite" metaphor doesn't work with a hypothetical "--target" option. But "recreate" does work: "borg, recreate (this archive) at (that location) with (these compression/chunker settings). Thanks."
  - "recreate" makes it clear that this tool works in similar ways and is essentially the same thing as "create" with the same semantics.

**Future TODOs:**
- Refactor tests using py.test fixtures
  -- would require porting ArchiverTestCase to py.test: many changes,
     this changeset is already borderline too large.
- Possibly add a --target option to not replace the source archive
  -- with the target possibly in another Repo
     (better than "cp" due to full integrity checking, and deduplication
      at the target)
- Detect and skip (unless --always-recompress) already recompressed chunks
  -- delayed until current PRs on borg.key APIs are decided
     #810 #789

Fixes #787 #686 #630 #70 (and probably some I overlooked)
Also see #757 and #770

---

~~Besides rewrite this PR includes two small bug fixes for the chunker, both of which should not have occurred in practice (but deserve fixing).~~, split into extra PR, merged into 1.0-maint (#819) and master (#821)

### CODE REVIEW

File: borg/testsuite/chunker.py [hunk 1]
```diff
@@ -29,3 +29,14 @@ def test_buzhash(self):
...
+    def test_small_reads(self):
+        class SmallReadFile:
+            input = b'a' * (20 + 1)
+
+            def read(self, nbytes):
+                self.input = self.input[:-1]
+                return self.input[:1]
+
+        reconstructed = b''.join(Chunker(0, *CHUNKER_PARAMS).chunkify(SmallReadFile()))
+        assert reconstructed == b'a' * 20
```
Comment (ThomasWaldmann) [inline borg/testsuite/chunker.py]: this commit looks like it should be a separate PR against 1.0-maint.

interesting find btw, I always wondered why the window size is 4095.

Comment (enkore) [inline borg/testsuite/chunker.py]: 1.0-maint is merged back into master at time, so I'll remove that commit from this PR, ok?

**Comment by ThomasWaldmann** [comment]
yes, i pull 1.0-maint into master regularly.

File: borg/archive.py [hunk 2]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+        def __init__(self):
+            self.stats = Statistics()
+
+    class Interrupted(Exception):
+        def __init__(self, metadata=None):
+            self.metadata = metadata or {}
+
+    @staticmethod
+    def is_temporary_archive(archive_name):
+        return archive_name.endswith('.rewrite') or archive_name.endswith('.rewrite.temp')
```
Comment (ThomasWaldmann) [inline borg/archive.py]: return archive_name.endswith(('.rewrite', '.rewrite.temp'))

File: borg/archiver.py [hunk 3]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
                                type=location_validator(archive=False),
                                help='path to the repository to be upgraded')
 
+        rewrite_epilog = textwrap.dedent("""
+        Rewrites the contents of existing archives.
+
+        --exclude, --exclude-from and PATH have the exact same semantics
+        as in borg create, this means if a PATH is specified the
+        rewrite includes that path and nothing else: PATH does *not* restrict
+        the rewrite to a path.
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: this phrase is ambiguous.

i guess you did not mean:
this means if a PATH is specified the rewrite includes ( that path and nothing else ).

but rather:
this means if a PATH is specified the rewrite includes that path and it nothing mean anything else.

File: borg/archiver.py [hunk 4]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
+
+        --compression recompresses all chunks. Due to how Borg stores compressed size
+        information this might display incorrect information for archives that were not
+        rewritten at the same time.
+        There is no risk of data loss by this. Use --force to recompress chunks already
+        using the specified compression algorithm.
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: already compressed using

File: borg/archiver.py [hunk 5]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
 
+        rewrite_epilog = textwrap.dedent("""
+        Rewrites the contents of existing archives.
+
+        --exclude, --exclude-from and PATH have the exact same semantics
+        as in borg create, this means if a PATH is specified the
+        rewrite includes that path and nothing else: PATH does *not* restrict
+        the rewrite to a path.
+
+        --compression recompresses all chunks. Due to how Borg stores compressed size
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: hmm, "recompresses all chunks" is not quite true as you mention in the next sentences.
maybe say "makes all chunks being compressed by the given method"?

File: borg/archiver.py [hunk 6]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
+        information this might display incorrect information for archives that were not
+        rewritten at the same time.
+        There is no risk of data loss by this. Use --force to recompress chunks already
+        using the specified compression algorithm.
+
+        --chunker-params will re-chunk all files in the archive, this can be
+        used to have upgraded Borg 0.xx or Attic archives deduplicate with
+        Borg 1.x archives.
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: great :)

File: borg/archiver.py [hunk 7]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
+
+        borg rewrite is signal safe. Send either SIGINT (Ctrl-C on most terminals) or
+        SIGTERM to request termination.
+
+        Use the *exact same* command line to resume the operation later - changing excludes
+        or paths will lead to inconsistencies (changed excludes will only apply to newly
+        processed files/dirs). Changing compression leads to incorrect size information
+        (which does not cause any data loss, but can be misleading).
+
+        USE WITH CAUTION. Permanent data loss by specifying incorrect patterns is possible.
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: ... incorrect PATHs or exclude patterns ...?

File: borg/archiver.py [hunk 8]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
+
+        Note: The archive under rewrite is only removed after the operation completes. The
+              archive that is built during the rewrite exists at the same time at
+              <ARCHIVE>.rewrite.
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: btw, "rewrite" sounds a bit like borg somehow writes into same archive, modifying it.
would "recreate" be clearer - it creates a new archive using existing data from an old one?

File: borg/archiver.py [hunk 9]
```diff
@@ -1514,6 +1553,102 @@ def build_parser(self, args=None, prog=None):
...
+                                          formatter_class=argparse.RawDescriptionHelpFormatter,
+                                          help='create backup')
+        subparser.set_defaults(func=self.do_rewrite)
+        subparser.add_argument('--list', dest='output_list',
+                               action='store_true', default=False,
+                               help='output verbose list of items (files, dirs, ...)')
+        subparser.add_argument('-p', '--progress', dest='progress',
+                               action='store_true', default=False,
+                               help='show progress display while rewriting archives')
+        subparser.add_argument('-f', '--force', dest='force_recompress',
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: -f or --force might be a bit too generic here?
imagine if we want to also force some other stuff (but not this one).
--force-recompress maybe?

File: borg/archive.py [hunk 10]
```diff
@@ -231,7 +233,7 @@ def add_item(self, item):
         if self.show_progress:
             self.stats.show_progress(item=item, dt=0.2)
         self.items_buffer.add(item)
-        if time.time() - self.last_checkpoint > self.checkpoint_interval:
+        if self.checkpoint_interval and time.time() - self.last_checkpoint > self.checkpoint_interval:
```
Comment (ThomasWaldmann) [inline borg/archive.py]: bugfix?

File: borg/testsuite/archiver.py [hunk 11]
```diff
@@ -274,7 +272,8 @@ def create_test_files(self):
             # same for newer ubuntu and centos.
             # if this is supported just on specific platform, platform should be checked first,
             # so that the test setup for all tests using it does not fail here always for others.
-            # xattr.setxattr(os.path.join(self.input_path, 'link1'), 'user.foo_symlink', b'bar_symlink', follow_symlinks=False)
+            # xattr.setxattr(os.path.join(self.input_path, 'link1'), 'user.foo_symlink', b'bar_symlink',
+            # follow_symlinks=False)
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: maybe stay away from reformatting the src while doing functional changes.
also, 80 < line length < 120 is pretty much ok for borg.

File: borg/testsuite/archiver.py [hunk 12]
```diff
@@ -1111,12 +1160,201 @@ def test_debug_put_get_delete_obj(self):
...
+
 
 @unittest.skipUnless('binary' in BORG_EXES, 'no borg.exe available')
 class ArchiverTestCaseBinary(ArchiverTestCase):
     EXE = 'borg.exe'
     FORK_DEFAULT = True
 
+    @unittest.skip('patches objects')
+    def test_rewrite_rechunkify_interrupt(self):
+        ...
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: pass?

or

"""something i could say here"""

File: borg/archiver.py (hunk 8 above)
Comment (enkore) [inline borg/archiver.py]: "rewrite" has this notion of "rewriting history", but the analogy doesn't really work if we consider `--target`. "borg recreate" would be clearer for those use cases, and it makes the link to "borg create" and how they both operate practically identical to each other.

File: borg/archive.py (hunk 10 above)
Comment (enkore) [inline borg/archive.py]: rewrite uses checkpoint_interval=0 (because it handles that explicitly). Previously this would mean something like "write a checkpoint for every item" (with a small timing dependency?).

File: borg/testsuite/archiver.py (hunk 12 above)
Comment (enkore) [inline borg/testsuite/archiver.py]: More like a `... there is a way we can somehow test this, and we probably should`. I haven't looked into that, though. If it's to complicated to test I'd put a `pass` in there.

File: borg/archiver.py (hunk 9 above)
Comment (enkore) [inline borg/archiver.py]: Yeah, I changed the TODO item to "--always-recompress", stuff like "--force" usually means something can/will go wrong, which isn't the case here.

File: borg/cache.py [hunk 13]
```diff
@@ -358,27 +358,29 @@ def legacy_cleanup():
...
         size = len(data)
-        if self.seen_chunk(id, size):
+        count, _, stored_csize = self.get_chunk(id, size)
+        if count and not check_csize:
             return self.chunk_incref(id, stats)
         data = self.key.encrypt(data)
         csize = len(data)
-        self.repository.put(id, data, wait=False)
-        self.chunks[id] = (1, size, csize)
+        if csize != stored_csize:
```
Comment (ThomasWaldmann) [inline borg/cache.py]: isn't that a bit backwards?

the real criteria should be whether we already have the wanted compression algorithm or not.

Comment (enkore) [inline borg/cache.py]: Since that logic would reside in ArchiveRewriter anyway, I'm changing this to an `overwrite` kwarg.

File: borg/archive.py [hunk 14]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+            if item_is_hardlink_master(item):
+                # Re-visit all of these items in the archive even when fast-forwarding to rebuild hardlink_masters
+                hardlink_masters[item[b'path']] = (item.get(b'chunks'), None)
+                continue
+            if not matcher.match(item[b'path']):
+                logger.info('x %s', remove_surrogates(item[b'path']))
+                continue
+            if resume_from:
+                # Fast forward to after the last processed file
+                if item[b'path'] == resume_from:
```
Comment (ThomasWaldmann) [inline borg/archive.py]: if A and B:

File: borg/archive.py [hunk 15]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+                chunks, new_source = hardlink_masters[item[b'source']]
+                if new_source is None:
+                    # First item to use this master, move the chunks
+                    item[b'chunks'] = chunks
+                    hardlink_masters[item[b'source']] = (None, item[b'path'])
+                    del item[b'source']
+                else:
+                    # Master was already moved, only update this item's source
+                    item[b'source'] = new_source
+            if not self.dry_run:
```
Comment (ThomasWaldmann) [inline borg/archive.py]: if dry_run, is it completely silent?

Comment (enkore) [inline borg/archive.py]: it would only print excluded stuff. I'll make it consistent with borg-create (`- file` for --list)

File: borg/archive.py [hunk 16]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+        if self.progress:
+            target.stats.show_progress(final=True)
+
+    def process_item(self, archive, target, item):
+        if b'chunks' in item:
+            self.process_chunks(archive, target, item)
+            target.stats.nfiles += 1
+        target.add_item(item)
+        if self.list:
+            logger.info('I %s', remove_surrogates(item[b'path']))
```
Comment (ThomasWaldmann) [inline borg/archive.py]: uppercase is only for different states of regular files.
so maybe use a lowercase i if you do not differentiate.

Comment (enkore) [inline borg/archive.py]: I'll make it consistent with what we already have in create; regular files are then always A (added).

File: borg/archive.py [hunk 17]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+            raise self.Interrupted
+
+    def process_chunks(self, archive, target, item):
+        if not self.recompress and not self.rechunkify:
+            for chunk in item[b'chunks']:
+                self.cache.chunk_incref(chunk[0], target.stats)
+            return
+        chunk_iterator = archive.pipeline.fetch_many([c[0] for c in item[b'chunks']])
+        if self.rechunkify:
+            file = ChunkIteratorFileWrapper(chunk_iterator)
```
Comment (ThomasWaldmann) [inline borg/archive.py]: file is a builtin.

File: borg/archive.py [hunk 18]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+            chunk_iterator = target.chunker.chunkify(file)
+        new_chunks = []
+        if target.rewrite_partial_chunks:
+            # No incref, create_target_or_resume already did that
+            new_chunks = target.rewrite_partial_chunks
+            target.rewrite_partial_chunks = None
+            for chunk in new_chunks:
+                if chunk[1] > 4096:
```
Comment (ThomasWaldmann) [inline borg/archive.py]: hm?

File: borg/archive.py (hunk 17 above)
Comment (enkore) [inline borg/archive.py]: file() was removed in py3 though

File: borg/archive.py (hunk 18 above)
Comment (enkore) [inline borg/archive.py]: small chunks are not put in the seen_chunks cache because they make little difference for performance but would still need the ~100 bytes/chunk or so in that cache.

File: borg/archive.py [hunk 19]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+                    self.seen_chunks.add(chunk_id)
+            if self.progress:
+                target.stats.show_progress(item=item, dt=0.2)
+            if self.interrupt:
+                raise self.Interrupted({
+                    'rewrite_partial_chunks': new_chunks,
+                })
+        if self.rechunkify:
+            assert file.exhausted
+        item[b'chunks'] = new_chunks
```
Comment (ThomasWaldmann) [inline borg/archive.py]: guess this method needs more docs/comments

File: borg/archive.py (hunk 18 above)
Comment (enkore) [inline borg/archive.py]: yeah, let's strip that. if those few MB blow the memory up on a machine it's too close to fail anyway. if that turns out to be a real issue in the field we can add a --save-mem(ory) in the spirit of --save-space; work much slower but use less space/mem.

File: borg/archive.py (hunk 17 above)
Comment (ThomasWaldmann) [inline borg/archive.py]: oh, didn't notice that yet (always use with open() as f). :)

File: borg/archive.py [hunk 20]
```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
...
+            logger.warning('Current fingerprint: %s', archive.fpr)
+            old_target.delete(Statistics(), progress=self.progress)
+            return None, None  # can't resume
+        if resume_args != sys.argv[1:]:
+            logger.warning('Command line changed, this might lead to inconsistencies')
+            logger.warning('Saved:   %s', repr(resume_args))
+            logger.warning('Current: %s', repr(sys.argv[1:]))
+        target_name += '.temp'
+        if target_name in self.manifest.archives:
+            logger.warning('Found temporary replay-archive from earlier resume, deleting')
```
Comment (ThomasWaldmann) [inline borg/archive.py]: replay archive? (in english usually no dashes)

Comment (enkore) [inline borg/archive.py]: this path was unnecessary, I removed it. (the archive is never *saved* with that name)

**Comment by enkore** [comment]
Renaming and squashing aside this changeset should be basically done.

File: borg/archive.py [hunk 21]
```diff
@@ -263,8 +265,9 @@ def save(self, name=None, timestamp=None):
             'time': start.isoformat(),
             'time_end': end.isoformat(),
             'chunker_params': self.chunker_params,
-        })
-        data = msgpack.packb(metadata, unicode_errors='surrogateescape')
+        }
+        metadata.update(additional_metadata or {})
+        data = msgpack.packb(StableDict(metadata), unicode_errors='surrogateescape')
```
Comment (enkore) [inline borg/archive.py]: This part likely merge-conflicts with #847. I'd say we merge #847 first and then I'll add "--comment" to this PR's "borg recreate" command, replacing the "borg comment" command from #847.

Comment (Ape) [inline borg/archive.py]: That's a good idea. I'm not familiar with the details of this pull request so it's better that you handle the conflicts.

Comment (enkore) [inline borg/archive.py]: All done :)

File: borg/archiver.py [hunk 22]
```diff
@@ -823,6 +811,50 @@ def do_upgrade(self, args):
...
         return self.exit_code
 
+    @with_repository(cache=True, exclusive=True)
+    def do_recreate(self, args, repository, manifest, key, cache):
+        """Re-create archives"""
+        def interrupt(signal_num, stack_frame):
+            if recreater.interrupt:
+                print("Received signal, again. I'm not deaf.\n", file=sys.stderr)
+            else:
+                print("Received signal, will exit cleanly.\n", file=sys.stderr)
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: why do you have a lf at the end of the message?

File: borg/archiver.py [hunk 23]
```diff
@@ -823,6 +811,50 @@ def do_upgrade(self, args):
...
+        recreater = ArchiveRecreater(repository, manifest, key, cache, matcher,
+                                     exclude_caches=args.exclude_caches, exclude_if_present=args.exclude_if_present,
+                                     keep_tag_files=args.keep_tag_files,
+                                     compression=args.compression, chunker_params=args.chunker_params,
+                                     progress=args.progress, stats=args.stats,
+                                     file_status_printer=self.print_file_status,
+                                     dry_run=args.dry_run)
+
+        signal.signal(signal.SIGTERM, interrupt)
+        signal.signal(signal.SIGINT, interrupt)
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: did you check that the repo can be brought into a good state even if another signal was used? e.g. sigkill?

File: borg/helpers.py [hunk 24]
```diff
@@ -1272,3 +1278,74 @@ def format_time(self, key, item):
...
+        return 'd'
+    elif stat.S_ISBLK(mode):
+        return 'b'
+    elif stat.S_ISCHR(mode):
+        return 'c'
+    elif stat.S_ISLNK(mode):
+        return 's'
+    elif stat.S_ISFIFO(mode):
+        return 'f'
+    return '?'
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: is ISLNK missing here?

File: borg/testsuite/archiver.py [hunk 25]
```diff
@@ -880,6 +926,13 @@ def test_file_status_excluded(self):
         self.assert_in("U input/file1", output)
         self.assert_in("x input/file2", output)
 
+    def test_create_delete_inbetween(self):
+        self.create_test_files()
+        self.cmd('init', self.repository_location)
+        self.cmd('create', self.repository_location + '::test1', 'input')
+        self.cmd('delete', self.repository_location + '::test1')
+        self.cmd('create', self.repository_location + '::test2', 'input')
+
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: there is no assert in this test, what's the point?

File: borg/testsuite/archiver.py [hunk 26]
```diff
@@ -1149,6 +1202,176 @@ def raise_eof(*args):
...
+        with changedir('output'):
+            self.cmd('extract', self.repository_location + '::test2')
+            assert os.stat('input/dir1/hardlink').st_nlink == 4
+
+    def test_recreate_rechunkify(self):
+        with open(os.path.join(self.input_path, 'large_file'), 'wb') as fd:
+            fd.write(b'a' * 250)
+            fd.write(b'b' * 250)
+        self.cmd('init', self.repository_location)
+        self.cmd('create', '--chunker-params', '7,9,8,128', self.repository_location + '::test1', 'input')
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: as max chunksize == 2^9 == 512 and your file is just 500B long, we can't be sure it will yield more than 1 chunk.

File: borg/testsuite/archiver.py [hunk 27]
```diff
@@ -1149,6 +1202,176 @@ def raise_eof(*args):
...
+                                '--format', '{unique_chunks}'))
+
+    def test_recreate_recompress(self):
+        self.create_regular_file('compressible', size=10000)
+        self.cmd('init', self.repository_location)
+        self.cmd('create', self.repository_location + '::test', 'input')
+        list = self.cmd('list', self.repository_location + '::test', 'input/compressible',
+                        '--format', '{size} {csize}')
+        size, csize = map(int, list.split(' '))
+        assert csize >= size
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: maybe add a comment here: due to chunk metadata overhead

File: borg/testsuite/archiver.py [hunk 28]
```diff
@@ -1149,6 +1202,176 @@ def raise_eof(*args):
...
+        assert num_chunks == unique_chunks
+        self.cmd('recreate', self.repository_location, '--chunker-params', 'default')
+        # test1 and test2 do deduplicate after recreate
+        assert not int(self.cmd('list', self.repository_location + '::test1', 'input/large_file',
+                                '--format', '{unique_chunks}'))
+
+    def test_recreate_recompress(self):
+        self.create_regular_file('compressible', size=10000)
+        self.cmd('init', self.repository_location)
+        self.cmd('create', self.repository_location + '::test', 'input')
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: maybe explicitly give --compression none here.

File: borg/testsuite/archiver.py [hunk 29]
```diff
@@ -1149,6 +1202,176 @@ def raise_eof(*args):
...
+        list = self.cmd('list', self.repository_location + '::test', 'input/compressible',
+                        '--format', '{size} {csize}')
+        size, csize = map(int, list.split(' '))
+        assert csize >= size
+        self.cmd('recreate', self.repository_location, '-C', 'lz4')
+        list = self.cmd('list', self.repository_location + '::test', 'input/compressible',
+                        '--format', '{size} {csize}')
+        size, csize = map(int, list.split(' '))
+        assert csize < size
+
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: also: assert size_after == size_before and content_after == content_before.

File: borg/testsuite/archiver.py [hunk 30]
```diff
@@ -1149,6 +1202,176 @@ def raise_eof(*args):
...
+            return real_add_chunk(*args, **kwargs)
+        return add_chunk
+
+    def test_recreate_rechunkify_interrupt(self):
+        self.create_regular_file('file1', size=1024 * 80)
+        self.cmd('init', self.repository_location)
+        self.cmd('create', self.repository_location + '::test', 'input')
+        archive_before = self.cmd('list', self.repository_location + '::test', '--format', '{sha512}')
+        with patch.object(Cache, 'add_chunk', self._test_recreate_chunker_interrupt_patch()):
+            self.cmd('recreate', '-p', '--chunker-params', '16,18,17,4095', self.repository_location)
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: with 2^18 max chunksize, 80KB might just yield 1 chunk.

File: docs/usage/recreate.rst.inc [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+    PATH
+        paths to recreate; patterns are supported
+
+optional arguments
+    ``--list``
+        | output verbose list of items (files, dirs, ...)
+    ``--filter STATUSCHARS``
+        | only display items with the given status characters
+    ``-p``, ``--progress``
+        | show progress display while rewriting archives
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: recreating

File: borg/archiver.py (hunk 22 above)
Comment (enkore) [inline borg/archiver.py]: Ah, they're on the wrong end. The LF must be before the message: If a progress line is around this will force this status message into the next line. Otherwise it will just partially overwrite the progress line.

File: docs/usage/recreate.rst.inc [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+        | specify the chunker parameters (or "default").
+
+Description
+~~~~~~~~~~~
+
+Recreate the contents of existing archives.
+
+--exclude, --exclude-from and PATH have the exact same semantics
+as in "borg create". If a PATH is specified the resulting archive
+will only contain files under PATH.
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: as one can give multiple PATHs, the "only" (referring to a single PATH) somehow doesn't make sense.

File: borg/archiver.py (hunk 23 above)
Comment (enkore) [inline borg/archiver.py]: Yes. If SIGKILL is used locks will get left behind and the indices are likely outdated and need rebuilding. The automatic segment replay will discard uncommitted changes.

File: docs/usage/recreate.rst.inc [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+Due to how Borg stores compressed size information this might display
+incorrect information for archives that were not rewritten at the same time.
+There is no risk of data loss by this.
+
+--chunker-params will re-chunk all files in the archive, this can be
+used to have upgraded Borg 0.xx or Attic archives deduplicate with
+Borg 1.x archives.
+
+borg recreate is signal safe. Send either SIGINT (Ctrl-C on most terminals) or
+SIGTERM to request termination.
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: a word about other signals maybe?

File: docs/usage/recreate.rst.inc [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+
+Use the *exact same* command line to resume the operation later - changing excludes
+or paths will lead to inconsistencies (changed excludes will only apply to newly
+processed files/dirs). Changing compression leads to incorrect size information
+(which does not cause any data loss, but can be misleading).
+Changing chunker params between invocations might lead to data loss.
+
+USE WITH CAUTION.
+Permanent data loss by specifying incorrect patterns or PATHS is possible.
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: true, but could be misunderstood (the data loss is due to admin failure, not due to borg).
maybe we could say:
Depending on the PATHs and patterns given, recreate can be used to permanently delete files from archives.

File: docs/usage/recreate.rst.inc [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+When in doubt, use "--dry-run --verbose --list" to see how patterns/PATHS are
+interpreted.
+
+The archive being recreated is only removed after the operation completes. The
+archive that is built during the operation exists at the same time at
+"<ARCHIVE>.recreate". The new archive will have a different archive ID.
+
+When rechunking space usage can be substantial, expect at least the entire
+deduplicated size of the archives using the older chunker params.
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: older -> original / previous?

File: docs/usage/recreate.rst.inc (hunk 33 above)
Comment (enkore) [inline docs/usage/recreate.rst.inc]: Hmm, recreate isn't different from other commands in how signals affect it except those mentioned. I.e. stale locks, inconsistent indices and partial transactions.

**Comment by enkore** [comment]
Good catches. While working with the tests I also found+fixed a small bug (nothing dramatic, really, just some data loss :D)

File: borg/testsuite/archiver.py [hunk 36]
```diff
@@ -1249,16 +1242,17 @@ def test_recreate_rechunkify(self):
...
         self.create_regular_file('compressible', size=10000)
         self.cmd('init', self.repository_location)
-        self.cmd('create', self.repository_location + '::test', 'input')
-        list = self.cmd('list', self.repository_location + '::test', 'input/compressible',
-                        '--format', '{size} {csize}')
-        size, csize = map(int, list.split(' '))
-        assert csize >= size
+        self.cmd('create', self.repository_location + '::test', 'input', '-C', 'none')
+        file_list = self.cmd('list', self.repository_location + '::test', 'input/compressible',
+                             '--format', '{size} {csize} {sha256}')
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: hah! :)

File: borg/archive.py [hunk 37]
```diff
@@ -1221,7 +1222,19 @@ def try_resume(self, archive, target_name):
         if self.progress:
             old_target.stats.show_progress(final=True)
         target.recreate_partial_chunks = old_target.metadata.get(b'recreate_partial_chunks', [])
-        for chunk_id, _, _ in target.recreate_partial_chunks:
+        for chunk_id, size, csize in target.recreate_partial_chunks:
+            if not self.cache.seen_chunk(chunk_id):
+                try:
+                    # Repository has __contains__, RemoteRepository doesn't
```
Comment (ThomasWaldmann) [inline borg/archive.py]: hmm, maybe we should remove __contains__ then?

**Comment by ThomasWaldmann** [comment]
as this feature is rather complex and initially not much practically tested, should it be flagged as "experimental" and have a similar warning like "borg check --repair"?

File: borg/archive.py (hunk 37 above)
Comment (enkore) [inline borg/archive.py]: Also doesn't seem to have any obvious usages

**Comment by RonnyPfannschmidt** [comment]
why is recreate a single command?
